utils/merge-rulesets.py

```diff
@@ -1,105 +1,97 @@
-    rulefile = re.sub(r"</ruleset>\s*", r"</ruleset>\n", rulefile)
-    rulefile = re.sub(r"\s*(/>|<ruleset)", r"\1", rulefile)
-    rulefile = re.sub(r"<test.+?/>", r"", rulefile)
-    return rulefile
-
-library = open(rulesets_fn,"w")
-
-try:
-  commit_id = os.environ["GIT_COMMIT_ID"]
-  library.write('<rulesetlibrary gitcommitid="%s">' % commit_id)
-except:
-  # Chromium
-  library.write('<rulesetlibrary>')
-
-# Include the filename.xml as the "f" attribute
-print("Removing whitespaces and comments...")
-
-for rfile in sorted(xml_ruleset_files):
-  ruleset = open(rfile).read()
-  fn = os.path.basename(rfile)
-  ruleset = ruleset.replace("<ruleset", '<ruleset f="%s"' % fn, 1)
-  library.write(clean_up(ruleset))
-library.write("</rulesetlibrary>\n")
-library.close()
-
-try:
-  if 0 == call(["xmllint", "--noout", rulesets_fn]):
-    print(rulesets_fn, "passed XML validity test.")
-  else:
-    print("ERROR:", rulesets_fn, "failed XML validity test!")
-    sys.exit(1)
-except OSError as e:
-  if "No such file or directory" not in traceback.format_exc():
-    raise
-  print("WARNING: xmllint not present; validation of", rulesets_fn, " skipped.")
-
-# We make default.rulesets at build time, but it shouldn't have a variable
-# timestamp
-call(["touch", "-r", "src/install.rdf", rulesets_fn])
+# output filename, pointed to the merged ruleset
+ofn = os.path.join(args.source_dir, "default.rulesets")
+
+# XML Ruleset Files
+files = map(normalize, glob.glob(os.path.join(args.source_dir, "*.xml")))
+
+# Under git bash, sed -i issues errors and sets the file "read-only".
+if os.path.isfile(ofn):
+	os.system("chmod u+w " + ofn)
+
+# Library (JSON Object)
+library = []
+
+# Parse XML ruleset and construct JSON library
+print(" * Parsing XML ruleset and constructing JSON library...")
+for filename in sorted(files):
+	tree = xml.etree.ElementTree.parse(filename)
+	root = tree.getroot()
+	
+	ruleset = {}
+
+	for attr in root.attrib:
+		ruleset[attr] = root.attrib[attr]
+	
+	for child in root:
+		if child.tag in ["target", "rule", "securecookie", "exclusion"]:
+			if child.tag not in ruleset:
+				ruleset[child.tag] = []
+		else:
+			continue
+
+		if child.tag == "target":
+			ruleset["target"].append(child.attrib["host"])
+
+		elif child.tag == "rule":
+			ru = {}
+			ru["from"] = child.attrib["from"]
+			ru["to"] = child.attrib["to"]
+
+			ruleset["rule"].append(ru)
+
+		elif child.tag == "securecookie":
+			sc = {}
+			sc["host"] = child.attrib["host"]
+			sc["name"] = child.attrib["name"]
+
+			ruleset["securecookie"].append(sc)
+
+		elif child.tag == "exclusion":
+			ruleset["exclusion"].append(child.attrib["pattern"])
+
+	library.append(ruleset);
+
+# Write to default.rulesets
+print(" * Writing JSON library to %s" % ofn)
+outfile = open(ofn, "w")
+outfile.write(json.dumps(library))
+outfile.close()
+
+# We make default.rulesets at build time, 
+# but it shouldn't have a variable timestamp
+subprocess.call(["touch", "-r", "src/install.rdf", ofn])
```

Comment:
This is no longer needed, `install.rdf` is going away with WebExtensions transition.

Reply:
This line is intended to remove the timestamp of default.rulesets. If src/install.rdf is gone, we will need another for timestamp reference. Let's see if Hainish would like to keep this line and suggest another file for timestamp reference.

P.S. Python because this will introduce fewer changes, I assume this will speed up the review process. thanks.

Reply:
@cschanaj Still all Python scripts will be replaced by Node scripts soon.